torch/csrc/distributed/autograd/utils.cpp

```diff
@@ -123,21 +123,27 @@ std::shared_ptr<FutureMessage> sendMessageWithAutograd(
       MessageType::FORWARD_AUTOGRAD_REQ,
       forceGradRecording);
 
-  auto fut = agent.send(dst, std::move(msg));
   if (rf != nullptr) {
     // save the local threadId so that end() callbacks can be correctly invoked
     // from a different thread.
     rf->setThreadId();
-    // Add a callback to
-    // the future that captures the RecordFunction to persist it for the
-    // lifetime of the future. When the future is completed, this will run the
-    // end() callbacks associated with the RecordFunction, so that async RPCs
-    // can be profiled correctly.
+    // This must be set synchronously in the same thread to avoid issues with
+    // nested scopes.
+    rf->resetThreadLocalFunc();
```

Comment:
I moved `rf->setThreadId()` and `rf->resetThreadLocalFunc()` to be before creating the future, since if we do this after, there is a small chance that the future could actually complete before we set the necessary information on the record function and return back to the user.